File: FinalProject/2_GraphAndTree/Prob14/AdjMatrix.cpp
#include <bits/stdc++.h>
using namespace std;

class Graph {
private:
    int n;
    vector<vector<int>>adj;

public:
    Graph(int n) : n(n), adj(n, vector<int>(n, 0)) {}

    void add_edge(int v, int w) {
        adj[v][w]=1;
    }

    void del_edge(int v, int w) {
        adj[v][w]=0;
    }

    vector<pair<int,int>> edges() {
        vector<pair<int,int>> res;
        for (int v=0; v<n; ++v) {
            for (int w=0; w<n; ++w) {
                if (adj[v][w]==1) {
                    res.emplace_back(v, w);
                }
            }
        }
        return res;
    }

    vector<pair<int,int>> incoming(int v) {
        vector<pair<int,int>> res;
        for (int u=0; u<n; ++u) {
            if (adj[u][v]==1) {
                res.emplace_back(u, v);
            }
        }
        return res;
    }

    vector<pair<int,int>> outgoing(int v) {
        vector<pair<int,int>> res;
        for (int w=0; w<n; ++w) {
            if (adj[v][w]==1) {
                res.emplace_back(v, w);
            }
        }
        return res;
    }

    // Trả về đỉnh nguồn của cạnh (v,w) nếu tồn tại, -1 nếu không
    int source(int v, int w) {
        if (adj[v][w]==1) return v;
        return -1;
    }

    // Trả về đỉnh đích của cạnh (v,w) nếu tồn tại, -1 nếu không
    int target(int v, int w) {
        if (adj[v][w]==1) return w;
        return -1;
    }

    void print() {
        cout << "Adjacency Matrix:\n";
        for (const auto& row : adj) {
            for (int val : row) {
                cout<<val<<" ";
            }
            cout<<"\n";
        }
    }
};

// Hàm hỗ trợ in vector các cặp (edges)
void print_edges(const vector<pair<int,int>>& edges) {
    cout<<"[";
    for (size_t i=0; i<edges.size(); ++i) {
        cout<<"("<<edges[i].first<<","<<edges[i].second<<")";
        if (i+1<edges.size()) cout<< ", ";
    }
    cout<<"]\n";
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    Graph G(5);
    G.add_edge(0, 1);
    G.add_edge(1, 2);
    G.add_edge(2, 3);
    G.add_edge(3, 4);
    G.add_edge(4, 0);

    G.print();

    cout<<"\nEdges: ";
    print_edges(G.edges());

    cout<<"\nIncoming to 3: ";
    print_edges(G.incoming(3));

    cout<<"\nOutgoing from 1: ";
    print_edges(G.outgoing(1));

    cout<<"\nSource of edge (2,3): "<<G.source(2, 3)<<"\n";
    cout<<"\nTarget of edge (2,3): "<<G.target(2, 3)<<"\n";

    G.del_edge(2, 3);
    cout<<"\nAfter deleting edge (2,3):"<<"\n";
    G.print();

    return 0;
}

File: FinalProject/2_GraphAndTree/Prob14/AdjMatrix.py
class Graph:
    def __init__(self, n):
        self.n = n
        self.adj = [[0]*n for _ in range(n)]

    def add_edge(self, v, w):
        self.adj[v][w] = 1

    def del_edge(self, v, w):
        self.adj[v][w] = 0

    def edges(self):
        # Trả về danh sách tất cả các cạnh dưới dạng cặp đỉnh (v, w)
        return [(v, w) for v in range(self.n) for w in range(self.n) if self.adj[v][w] == 1]

    def incoming(self, v):
        # Trả về danh sách các cạnh đi vào đỉnh v: các cặp (u, v)
        return [(u, v) for u in range(self.n) if self.adj[u][v] == 1]

    def outgoing(self, v):
        # Trả về danh sách các cạnh đi ra từ đỉnh v: các cặp (v, w)
        return [(v, w) for w in range(self.n) if self.adj[v][w] == 1]

    def source(self, v, w):
        # Trả về đỉnh nguồn của cạnh (v, w) nếu cạnh tồn tại
        if self.adj[v][w] == 1:
            return v
        else:
            return None

    def target(self, v, w):
        # Trả về đỉnh đích của cạnh (v, w) nếu cạnh tồn tại
        if self.adj[v][w] == 1:
            return w
        else:
            return None

    def __str__(self):
        res = "Adjacency Matrix:\n"
        for row in self.adj:
            res += " ".join(map(str, row)) + "\n"
        return res


# Ví dụ sử dụng
if __name__ == "__main__":
    G = Graph(5)
    G.add_edge(0, 1)
    G.add_edge(1, 2)
    G.add_edge(2, 3)
    G.add_edge(3, 4)
    G.add_edge(4, 0)

    print(G)

    print("Edges:", G.edges())
    print("Incoming to 3:", G.incoming(3))
    print("Outgoing from 1:", G.outgoing(1))

    print("Source of edge (2, 3):", G.source(2, 3))
    print("Target of edge (2, 3):", G.target(2, 3))

    G.del_edge(2, 3)
    print("After deleting edge (2, 3):")
    print(G)
